src/api/ui/views/task.tsx

```diff
@@ -0,0 +1,227 @@
+import { MarkdownListItem, MarkdownTaskItem } from "index/types/markdown";
+import { DefaultListElement, ListState } from "api/ui/views/list";
+import { useIndexUpdates, useStableCallback } from "ui/hooks";
+import { Fragment } from "preact";
+import { APP_CONTEXT, DATACORE_CONTEXT } from "ui/markdown";
+import { JSXInternal } from "preact/src/jsx";
+import { Dispatch, useContext, useMemo, useRef, useState } from "preact/hooks";
+import { rewriteTask, setTaskCompletion } from "ui/utils";
+import { Literal, Literals } from "expression/literal";
+import {
+    EditableAction,
+    EditableListField,
+    EditableState,
+    TextEditable,
+    useEditableDispatch,
+} from "ui/fields/editable";
+import { setInlineField } from "index/import/inline-field";
+import { BaseFieldProps } from "ui/fields/common-props";
+import { Field } from "expression/field";
+import { DateTime } from "luxon";
+import "styles/lists.css";
+
+export interface TaskProps extends ListState<MarkdownTaskItem | MarkdownListItem> {
+    /** task states to cycle through, if specified */
+    additionalStates?: string[];
+    /** fields to display under each item in this task list */
+    displayedFields?: (BaseFieldProps<Literal> & { key: string })[];
+}
+
+export function TaskList({
+    rows: items,
+    additionalStates: states,
+    renderer: listRenderer = (item, index) => <DefaultListElement element={item} />,
+    ...rest
+}: TaskProps) {
+    const content = useMemo(() => {
+        return (
+            <ul className="datacore contains-task-list">
+                {items?.map((item, ind) =>
+                    item instanceof MarkdownTaskItem ? (
+                        <Task state={{ ...rest, additionalStates: states }} item={item} />
+                    ) : (
+                        listRenderer(item, ind)
+                    )
+                )}
+            </ul>
+        );
+    }, [items, states]);
+    return <Fragment>{!!items && content}</Fragment>;
+}
+
+export function Task({ item, state: props }: { item: MarkdownTaskItem; state: TaskProps }) {
+    const app = useContext(APP_CONTEXT);
+    const core = useContext(DATACORE_CONTEXT);
+    const { settings } = core;
+    const states = [" ", ...(props.additionalStates || []), "x"];
+    const nextState = useMemo(() => {
```

Comment:
This doesn't need to be a memoized value - you can just extract this to a function you call whenever the checkmark is checked. The memoization isn't particularly useful.